Reduce use of `const char*` in <wtf/DateMath.h>

cdumez opened the pull request:
#### 511e5fff44f9949be51d903a8d1f9c851d2864e0
<pre>
Reduce use of `const char*` in &lt;wtf/DateMath.h&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=273409">https://bugs.webkit.org/show_bug.cgi?id=273409</a>

Reviewed by Darin Adler.

Reduce use of `const char*` in &lt;wtf/DateMath.h&gt; by leveraging std::span.

* Source/JavaScriptCore/runtime/JSDateMath.cpp:
(JSC::DateCache::parseDate):
* Source/WTF/wtf/DateMath.cpp:
(WTF::skipSpacesAndComments):
(WTF::findMonth):
(WTF::parseInt):
(WTF::parseLong):
(WTF::parseDate):
(WTF::parseDateFromNullTerminatedCharacters): Deleted.
* Source/WTF/wtf/DateMath.h:
* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::parseHTTPDate):
* Source/WebCore/platform/network/curl/CookieUtil.cpp:
(WebCore::CookieUtil::parseExpiresMS):
(WebCore::CookieUtil::parseCookieAttributes):

Canonical link: <a href="https://commits.webkit.org/278153@main">https://commits.webkit.org/278153@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/305c5c2945b03c8ad44ca31670cabac1714f821f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49651 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52700 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52893 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/327 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51956 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34958 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26556 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51751 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26470 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21638 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23914 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43951 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8022 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/42966 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44455 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54470 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/49146 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24738 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20899 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47893 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26005 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26853 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/56633 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7145 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/25730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11637 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/JSDateMath.cpp:
```diff
@@ -451,11 +451,11 @@ double DateCache::parseDate(JSGlobalObject* globalObject, VM& vm, const String&
+    auto parseDateImpl = [this] (std::span<const LChar> dateString) {
         bool isLocalTime;
-        double value = WTF::parseES5DateFromNullTerminatedCharacters(dateString, isLocalTime);
+        double value = WTF::parseES5DateFromNullTerminatedCharacters(reinterpret_cast<const char*>(dateString.data()), isLocalTime);
```
This seems risky. What guarantees that `dateString.data()` is null-terminated?

Also makes my `byteCast` job harder by adding **new** call sites for it, which I will might miss when I merge! (Just kidding on this comment, though.)

cdumez replied:
Oh, I was trying to not update parseES5DateFromNullTerminatedCharacters() in this patch to reduce patch size but I may indeed have introduced a bug as a result. I may have to bite the bullet and update parseES5DateFromNullTerminatedCharacters() to take a span too.

cdumez replied:
Fixed. I fixed at the call site so I don't have to convert parseES5DateFromNullTerminatedCharacters() in this PR.

darinadler commented on Source/WTF/wtf/DateMath.cpp:
```diff
@@ -479,6 +492,17 @@ static bool parseLong(const char* string, char** stopPosition, int base, long* r
     return true;
 }
 
+static bool parseLong(std::span<const LChar>& string, int base, long* result)
```
So sad since `long` and `int` are the same width on virtually every platform. Maybe this should be a template since the functions are identical except for the range check.

cdumez replied:
Fixed.

darinadler commented on Source/WTF/wtf/DateMath.cpp:
```diff
@@ -470,6 +470,19 @@ static bool parseInt(const char* string, char** stopPosition, int base, int* res
+    if (string.data() == reinterpret_cast<const LChar*>(stopPosition) || longResult <= std::numeric_limits<int>::min() || longResult >= std::numeric_limits<int>::max())
+        return false;
+    string = string.subspan(reinterpret_cast<const LChar*>(stopPosition) - string.data());
+    *result = static_cast<int>(longResult);
```
This `static_cast<int>` is not needed, has no effect since assigning already chops `long` to `int`.

cdumez replied:
Fixed.

cdumez committed:
Reduce use of `const char*` in <wtf/DateMath.h>
https://bugs.webkit.org/show_bug.cgi?id=273409

Reviewed by Darin Adler.

Reduce use of `const char*` in <wtf/DateMath.h> by leveraging std::span.

* Source/JavaScriptCore/runtime/JSDateMath.cpp:
(JSC::DateCache::parseDate):
* Source/WTF/wtf/DateMath.cpp:
(WTF::skipSpacesAndComments):
(WTF::findMonth):
(WTF::parseInt):
(WTF::parseLong):
(WTF::parseDate):
(WTF::parseDateFromNullTerminatedCharacters): Deleted.
* Source/WTF/wtf/DateMath.h:
* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::parseHTTPDate):
* Source/WebCore/platform/network/curl/CookieUtil.cpp:
(WebCore::CookieUtil::parseExpiresMS):
(WebCore::CookieUtil::parseCookieAttributes):

Canonical link: https://commits.webkit.org/278153@main
